src/Buffer.cpp

```diff
@@ -106,11 +106,8 @@ EXPORT RefCount &ref_count<BufferContents>(const BufferContents *p) {
 
 template<>
 EXPORT void destroy<BufferContents>(const BufferContents *p) {
-    // Free any device-side allocation
-    if (p->source_module.free_dev_buffer) {
-        int error = p->source_module.free_dev_buffer(NULL, const_cast<buffer_t *>(&p->buf));
-        user_assert(!error) << "Failed to free device buffer\n";
-    }
+    int error = p->source_module.dev_free(const_cast<buffer_t *>(&p->buf));
```

Comment:
Interesting question. destroy is typically a wrapper for delete, and delete
is perfectly happy to take a const pointer. But destroy is only ever called
from decref, which takes a non-const pointer.

Yeah I guess it should probably be non-const, but let's do that in a
different PR.

On Fri, Jan 16, 2015 at 9:27 AM, Dillon Sharlet <notifications@github.com>
wrote:

> In src/Buffer.cpp
> <https://github.com/halide/Halide/pull/630#discussion-diff-23094849>:
>
> > @@ -106,11 +106,8 @@ EXPORT RefCount &ref_count<BufferContents>(const BufferContents *p) {
> >
> >  template<>
> >  EXPORT void destroy<BufferContents>(const BufferContents *p) {
> > -    // Free any device-side allocation
> > -    if (p->source_module.free_dev_buffer) {
> > -        int error = p->source_module.free_dev_buffer(NULL, const_cast<buffer_t *>(&p->buf));
> > -        user_assert(!error) << "Failed to free device buffer\n";
> > -    }
> > +    int error = p->source_module.dev_free(const_cast<buffer_t *>(&p->buf));
>
> I know this isn't strictly part of this PR, but, it seems like destroy<>
> should take a non-const pointer.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/630/files#r23094849>.
>